# Current implementation
* Takes following inputs
    - query strains as genbank
    - reference metabolic model
        - the model file
        - corresponding genes and proteins
* Draft models for each strain are generated by
    - ortholog detection via blastp
        - not all genes may be annotated by prokka
        - use blastn to identify these
    - filtering low quality hits
    - removal of proteins and reactions from reference model
    - optimisation of model
* Growth simulation


# Planned features
* Preferentially take assembly graphs as input
    - to enable collection of deadend count
    - deadends likely important wrt to quality of model
* Reannotate inputs
    - using prodigal model trained on input set
        - computionally intensive for large inputs, optimise conditionally required
* Faster alternative to BLASTp
    - e.g. diamond
    - will need to demonstrate consistency between results
    - must also ensure that reasonable speed up is obtained
* Provide extensive information for draft genomes that fail to optimise
    - determine essential genes from reference and have these output somewhere
        - required only to be generated once (not per strain)


# Planned implementation
* Input as graphs or flat formats (FASTA, genbank, etc)
    - prefering graphs for additional QC
    - (deadends to be demonstrated as important through analysis later)
* QC of assemblies
* Annotation using same prodigal model
* Ortholog detection
* Draft model creatation
* Model optimisation
    - failing models set aside for manual investigation
    - large amount of info provide to help this process
* Growth simulation of specified media
    - should this run if some models failed? I think so


# Planned analysis
* Examine aspects of input assemblies that affect model quality
    - dataset with good quality genomes and biolog data
    - generate assemblies that have varying quality
        - deadends, contigs, completeness, etc
    - create models and test failures
        - quantify on basis of
            - number of genes removed
            - concordance of simulated growth to biolog data
